### DISCUSSION TITLE
how would I use the embed.FS that contains a cue.mod to bootstrap a cue.Context ?

### DISCUSSION THREAD

**Discussion by 4x4notfound** [cue-lang/cue#1117] [Q&A]
Building tools/cli where cue is used heavily to validating input. We'd like to be able to embed a cue.mod into the binary and boostrap a cue context with it.

**Comment by verdverm** [comment]
Have a look at `load.Instances` and the `load.Config.Overlay`, iterate over the embed.FS and insert into the overlay.

https://pkg.go.dev/cuelang.org/go@v0.4.0/cue/load

You'd feed the return into `ctx.BuildInstances`

You may have to do some fiddling with directory parameters as load will also look at the filesystem, so you won't want anything conflicting from there. You might prefix all of the filenames with an absolute (made up) path.

There is an open task to migrate to using `embed.FS` once Go 1.17 is released

**Reply by 4x4notfound** [reply]
Would this pattern be more efficient in 17?

related: https://github.com/cue-lang/cue/issues/1084

**Reply by verdverm** [reply]
The reason for Go 1.17 being a _requirement_ is that `cue` shares its supported versions with Go (i.e. the last two minor releases). Since `embed.FS` was introduced in Go 1.16, we need to wait until Go 1.17 in order to not break anything still on Go 1.15.

The general idea is to replace the overlay with `embed.FS`, so it would likely be more efficient, but is more of a restructuring of the loading process. There are a few other things that will be simplified or fixed by adopting `embed.FS`.

Related: https://github.com/cue-lang/cue/issues/607 https://github.com/cue-lang/cue/issues/851

**Reply by 4x4notfound** [reply]
What I've sorted out so far...

```
	config := load.Config{
		Overlay: make(map[string]load.Source),
	}
	fs.WalkDir(StaticFS, ".", func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if strings.Contains(path, ".cue") {
			f, err := ioutil.ReadFile(path)
			config.Overlay[path] = load.FromString(string(f))
			return err
		}
		if d != nil {
			return nil
		}
		return err
	})
	instances := load.Instances([]string{"kmdr"}, &config)
	kmdrInstance := instances[0]}
```

But not sure how to get a cuecontext from here

**Reply by 4x4notfound** [reply]
nvm, I can build a cue value with the instances using the cuecontexts

**Comment by myitcv** [comment]
@maurerbot thanks for raising this as a discussion. Moving good questions like this to GitHub discussions is really helping to improve the discoverability of CUE questions and answers.

To build on @verdverm's answer above, see also the example of https://github.com/play-with-go/preguide/blob/e5ab271ba2e92313c56754702035862b357bb48d/preguide.go#L71. 

To slightly caveat https://github.com/cue-lang/cue/discussions/1117#discussioncomment-1023148, we could have added Go 1.16-specific API to CUE using build tags, but it complicates things from a maintainer and user perspective. So far better to, as @verdverm suggested, keep it simple and introduce when we "drop" `<=1.15` support (in reality we will keep 1.15 in the build matrix and retain compatibility for longer, because some/many people do not/cannot closely adhere to the "latest two Go versions" policy).

**Reply by 4x4notfound** [reply]
I thought runtime is being deprecated?

**Reply by 4x4notfound** [reply]
here is the adjusted working code with `cuecontext`

```
	overlay := make(map[string]load.Source)
	err := fs.WalkDir(StaticFS, ".", func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d == nil || d.Type().IsDir() || !d.Type().IsRegular() {
			return nil
		}

		if strings.Contains(path, ".cue") {
			f, err := ioutil.ReadFile(path)
			if err != nil {
				return err
			}
			overlay[filepath.Join("/", path)] = load.FromBytes(f)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	config := &load.Config{
		Dir:     "/",
		Overlay: overlay,
	}
	instances := load.Instances([]string{"/static/kmdr"}, config)

	cueContext := cuecontext.New()
	defs := cueContext.BuildInstance(instances[0])
```

**Reply by myitcv** [reply]
> I thought runtime is being deprecated?

Ha ha! You busted me 😄 . That's simply a case of not having updated my own project to use the latest CUE (because if I had done, [`staticcheck`](https://github.com/play-with-go/preguide/blob/e5ab271ba2e92313c56754702035862b357bb48d/internal/ci/workflows.cue#L68-L69) would have caught this. Thank you for flagging!

**Reply by myitcv** [reply]
> here is the adjusted working code with `cuecontext`

FYI - this kind of example is exactly the kind of thing we want to surface via GitHub discussions. So thanks for posting this, @maurerbot. I would mark your own response as the answer if I were you 😄
